include/eve/detail/shuffle_v2/simd/common/shuffle_v2_driver.hpp

```diff
@@ -0,0 +1,600 @@
+    auto get_one_field = [&]<std::ptrdiff_t i>(index_t<i>)
+    { return kumi::map([](auto x) { return get<i>(x); }, xs); };
+
+    return [&]<std::size_t... field_i>(std::index_sequence<field_i...>) {
+      return kumi::tuple {get_one_field(index<(std::ptrdiff_t)field_i>)...};
+    }(std::make_index_sequence<std::tuple_size_v<T>> {});
+  }
+
+  template<typename Bundle, typename... Ts>
+  auto map_components(kumi::tuple<Ts...> result_per_field) const noexcept
+  {
+    auto field_results = kumi::transpose(result_per_field);
+
+    return kumi::map(
+        []<typename Field, typename... Fields>(kumi::tuple<Field, Fields...> field_res)
+        {
+          using N1 = typename Field::cardinal_type;
+          using T1 = typename Bundle::template rescale<N1>;
+          return T1 {field_res};
+        },
+        field_results);
+  }
+
+  template<typename NativeSelector, typename Pattern, typename G> struct recurse
+  {
+    NativeSelector selector;
+    recurse(NativeSelector selector, Pattern, G) : selector(selector) {}
+
+    EVE_FORCEINLINE
+    auto operator()(auto xs) const noexcept
+    {
+      return shuffle_v2_driver_bundle {}(selector, Pattern {}, G {}, xs);
+    }
+  };
+
+  template<typename NativeSelector,
+           std::ptrdiff_t G,
+           std::ptrdiff_t... I,
+           typename T,
+           typename... Ts>
+  EVE_FORCEINLINE auto
+  operator()(NativeSelector selector, pattern_t<I...> p, fixed<G> g, kumi::tuple<T, Ts...> xs) const
+  {
+    if constexpr( !product_simd_value<T> )
+    {
+      return shuffle_v2_driver_aggregation {}(selector, p, g, xs);
+    }
+    else
+    {
+      auto fields    = split_by_field(xs);
+      auto each_part = kumi::map(recurse {selector, p, g}, fields);
+
+      constexpr auto l = shuffle_v2_combined_l<decltype(each_part)>();
+      if constexpr( l() == -1 ) { return kumi::tuple {no_matching_shuffle_t {}, eve::index<-1>}; }
+      else
+      {
+        auto result_per_field = kumi::map([](auto x_l) { return get<0>(x_l); }, each_part);
+        auto shuffled         = map_components<T>(result_per_field);
+        return kumi::tuple {shuffled, l};
+      }
+    }
+  }
+};
+
+// start shuffling --------------------------------------
+
+template<typename NativeSelector, std::ptrdiff_t G, std::ptrdiff_t... I, typename T, typename... Ts>
+EVE_FORCEINLINE auto
+shuffle_v2_driver_start(NativeSelector        selector,
+                        pattern_t<I...>       p,
+                        fixed<G>              g,
+                        kumi::tuple<T, Ts...> xs)
+{
+  if constexpr( !eve::supports_simd ) return shuffle_emulated(p, g, xs);
+  else return shuffle_v2_driver_bundle {}(selector, p, g, xs);
+}
+
+// entry point -----------------------------------------
+
+template<typename T, typename... Ts>
+EVE_FORCEINLINE auto
+shuffle_v2_driver_construct_result(kumi::tuple<T, Ts...> shuffled)
+{
+  if constexpr( sizeof...(Ts) == 0 ) return get<0>(shuffled);
```

Comment:
this is quite a large file, any hope to split it up in smaller chunks ?

Reply:
I can move `simplify_plain_shuffle` into a separate file. Let me do that in a separate patch if you don't mind.